be/src/runtime/buffer_control_block.cpp

```diff
@@ -79,145 +80,162 @@ void GetResultBatchCtx::on_data(TFetchDataResult* t_result, int64_t packet_seq,
     delete this;
 }
 
+void GetResultBatchCtx::on_data(SerializeRes* res, int64_t packet_seq, bool eos) {
+    auto st = Status::OK();
+    cntl->response_attachment().swap(res->attachment);
+    result->set_packet_seq(packet_seq);
+    result->set_eos(eos);
+    st.to_protobuf(result->mutable_status());
+    done->Run();
+    delete this;
+}
+
 BufferControlBlock::BufferControlBlock(const TUniqueId& id, int buffer_size)
         : _fragment_id(id),
           _is_close(false),
           _is_cancelled(false),
-          _buffer_rows(0),
+          _buffer_bytes(0),
           _buffer_limit(buffer_size),
           _packet_num(0) {}
 
 BufferControlBlock::~BufferControlBlock() {
     cancel();
-
-    for (auto& iter : _batch_queue) {
-        delete iter;
-        iter = nullptr;
+    std::unique_ptr<SerializeRes> ser = nullptr;
+    while (_batch_queue.try_dequeue(ser)) {
+        ser.reset();
     }
+    _buffer_bytes = 0;
 }
 
 Status BufferControlBlock::init() {
     return Status::OK();
 }
 
-Status BufferControlBlock::add_batch(TFetchDataResult* result) {
-    std::unique_lock<std::mutex> l(_lock);
-
+Status BufferControlBlock::add_batch(TFetchDataResult* result, bool need_free) {
     if (_is_cancelled) {
         return Status::Cancelled("Cancelled BufferControlBlock::add_batch");
     }
-
-    int num_rows = result->result_batch.rows.size();
-
-    while ((!_batch_queue.empty() && (num_rows + _buffer_rows) > _buffer_limit) && !_is_cancelled) {
+    // serialize
+    auto status_or_value = _serialize_result(result);
```

Comment:
1. Why serialize data here?
2. Can we deserialize data after network?

Reply:
1. earily serialize before puting the queue, so the serialization cost is in computing threads, rather than rpc threads.
2. rpc destination will deserialize it.